experimental/internal/taxa/taxa.go

```diff
@@ -0,0 +1,217 @@
+// limitations under the License.
+
+// package taxa (plural of taxon, an element of a taxonomy) provides support for
+// classifying Protobuf syntax productions for use in the parser and in
+// diagnostics.
+//
+// The Subject enum is also used in the parser stack as a simple way to inform
+// recursive descent calls of what their caller is, since the What enum
+// represents "everything" the parser stack pushes around.
+package taxa
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/bufbuild/protocompile/internal/iter"
+)
+
+// Subject is a syntactic or semantic element within the grammar that can be
+// referred to within a diagnostic.
+type Subject int
+
+const (
+	Unknown Subject = iota
+	Unrecognized
+	TopLevel
+	EOF
+
+	Decl
+	Empty
+	Syntax
+	Edition
+	Package
+	Import
+	WeakImport
+	PublicImport
+	Extensions
+	Reserved
+	Body
+
+	Def
+	Message
+	Enum
+	Service
+	Extend
+	Oneof
+
+	Option
+	CustomOption
+
+	Field
+	EnumValue
+	Method
+
+	FieldTag
+	OptionValue
+
+	CompactOptions
+	MethodIns
+	MethodOuts
+
+	Path
+	ExtensionInPath
+
+	Expr
+	Range
+	Array
+	Dict
+	DictField
+
+	Type
+	TypeParams
+
+	Whitespace
+	Comment
+	Ident
+	String
+	Float
+	Int
+
+	Semicolon
+	Comma
+	Slash
+	Colon
+	Equals
+	Minus
+	Period
+
+	LParen
+	LBracket
+	LBrace
+	LAngle
+
+	RParen
+	RBracket
+	RBrace
+	RAngle
+
+	Parens
+	Brackets
+	Braces
+	Angles
+
+	KeywordSyntax
+	KeywordEdition
+	KeywordImport
+	KeywordWeak
+	KeywordPublic
+	KeywordPackage
+
+	KeywordOption
+	KeywordMessage
+	KeywordEnum
+	KeywordService
+	KeywordExtend
+	KeywordOneof
+
+	KeywordExtensions
+	KeywordReserved
+	KeywordTo
+	KeywordRPC
+	KeywordReturns
+
+	KeywordOptional
+	KeywordRepeated
+	KeywordRequired
+	KeywordGroup
+	KeywordStream
+
+	// Total is the total number of known [What] values.
+	Total int = iota
+)
+
+// In is a shorthand for the "in" preposition.
+func (s Subject) In() Place {
+	return Place{s, "in"}
+}
+
+// After is a shorthand for the "after" preposition.
+func (s Subject) After() Place {
+	return Place{s, "after"}
+}
+
+// Without is a shorthand for the "without" preposition.
+func (s Subject) Without() Place {
+	return Place{s, "without"}
+}
+
+// AsSet returns a singleton set containing this What.
+func (s Subject) AsSet() Set {
+	return NewSet(s)
+}
+
+// String implements [fmt.Stringer].
+func (s Subject) String() string {
+	if int(s) >= len(names) {
+		return names[0]
+	}
+	return names[s]
+}
+
+// All returns an iterator over all subjects.
+func All() iter.Seq[Subject] {
+	return func(yield func(Subject) bool) {
+		for i := 0; i < Total; i++ {
+			if !yield(Subject(i)) {
+				break
+			}
+		}
+	}
+}
+
+// GoString implements [fmt.GoStringer].
+//
+// This exists to get pretty output out of the assert package.
+func (s Subject) GoString() string {
+	if int(s) >= len(constNames) {
+		return strconv.Itoa(int(s))
+	}
+	return "what." + constNames[s]
+}
+
+// Place is a location within the grammar that can be referred to within a
+// diagnostic.
```

Comment:
I mentioned in a comment in other PR (the parser I think, before you extracted all of this to separate branches) that this whole package feels a bit over-engineered. This one in particular seems the lowest value -- do we really need an abstraction for `("... after %s", subj)`? It feels like we're building a DSL for English when just using English would be easier to read and write.

Reply:
It exists specifically because there are places where we want to pass in a place so that they can generate a generic diagnostic regardless of the place. This is more obvious once you get to the parser PR.